### ISSUE TITLE
[QUESTION] Is it possible to run LightGBM in useSingldeDatasetMode with all cores allocated until the end of fitting? 

### ISSUE THREAD

**Issue by fonhorst** [microsoft/SynapseML#1739]
Hello. I need to use parallel jobs in a single Spark application.
However, If I use lightgm with 'useSingleDataset=True' parameter it runs multiple tasks per executor (equal to a number of cores on it) and than all tasks except one will finish after data preparing.

![image_2022-11-22_23-44-25](https://user-images.githubusercontent.com/3002739/203530562-e0d00b7c-ea1c-4b47-aafb-8b36b33377c6.png)

The left one spawns several threads to execute training in parallel on the executor. But if there are another parallel job (or another instance of lightgbm), Spark would assign this parallel tasks on cores freed by the data preparation tasks. but still occupied by the lightgbm threads. It would lead to same cores being used by different tasks simultaneously which is not desirable.

I would like to know if there is a way to tell lightgbm not to free the cores until the training is ended?

**Comment by svotaw** [comment]
I'm not quite sure what you are asking. In useSingleDataSet=true mode, only 1 Task per executor is left after the data preparation stage, and all other Tasks on the same executor are completed.  LightGBM does not use any thread directly except for that 1 Task per executor (besides any OpenMP parallelism that happens inside the native LightGBM library itself). That's what makes useSingleDataSet=true more efficient... LightGBM is given only 1 thread per executor and there is less network time communicating between "machines".  Can you explain why you think threads are still occupied?

You can set things like numTasks to make it so that there's only one Task/partition per executor, but that doesn't limit LightGBM training from OpenMP parallelism.

And no, we don't have any way to "hold on" to Task threads that I know of. You could potentially use some direct LightGBM params or environment vars to affect OpenMP multithreading, but let's no go there unless that's really what you want to do.

LightGBM native library was not really designed to run in an external multi-threaded environment (e.g., your parallel jobs). It's designed to take advantage of all threads it can internally. We have a new execution mode that fixes some multi-threaded issues (to be released very soon), but it just allows multiple Tasks on the same executor to load the Dataset in parallel. It doesn't constrain OpenMP thread usage directly.

**Comment by shexuan** [comment]
> I'm not quite sure what you are asking. In useSingleDataSet=true mode, only 1 Task per executor is left after the data preparation stage, and all other Tasks on the same executor are completed. LightGBM does not use any thread directly except for that 1 Task per executor (besides any OpenMP parallelism that happens inside the native LightGBM library itself). That's what makes useSingleDataSet=true more efficient... LightGBM is given only 1 thread per executor and there is less network time communicating between "machines". Can you explain why you think threads are still occupied?
> 
> You can set things like numTasks to make it so that there's only one Task/partition per executor, but that doesn't limit LightGBM training from OpenMP parallelism.
> 
> And no, we don't have any way to "hold on" to Task threads that I know of. You could potentially use some direct LightGBM params or environment vars to affect OpenMP multithreading, but let's no go there unless that's really what you want to do.
> 
> LightGBM native library was not really designed to run in an external multi-threaded environment (e.g., your parallel jobs). It's designed to take advantage of all threads it can internally. We have a new execution mode that fixes some multi-threaded issues (to be released very soon), but it just allows multiple Tasks on the same executor to load the Dataset in parallel. It doesn't constrain OpenMP thread usage directly.

I have met the same problems, and according to your reply I reset my cluster resource 「200cores = 50executor x 4 core」to 「200cores = 200executor x 1core」greatly accelerate the traning time. But now I get stuck in other things https://github.com/microsoft/SynapseML/issues/1595. And I'm testing update the version from 0.10.0 to 0.10.2.

**Comment by shexuan** [comment]
> > I'm not quite sure what you are asking. In useSingleDataSet=true mode, only 1 Task per executor is left after the data preparation stage, and all other Tasks on the same executor are completed. LightGBM does not use any thread directly except for that 1 Task per executor (besides any OpenMP parallelism that happens inside the native LightGBM library itself). That's what makes useSingleDataSet=true more efficient... LightGBM is given only 1 thread per executor and there is less network time communicating between "machines". Can you explain why you think threads are still occupied?
> > You can set things like numTasks to make it so that there's only one Task/partition per executor, but that doesn't limit LightGBM training from OpenMP parallelism.
> > And no, we don't have any way to "hold on" to Task threads that I know of. You could potentially use some direct LightGBM params or environment vars to affect OpenMP multithreading, but let's no go there unless that's really what you want to do.
> > LightGBM native library was not really designed to run in an external multi-threaded environment (e.g., your parallel jobs). It's designed to take advantage of all threads it can internally. We have a new execution mode that fixes some multi-threaded issues (to be released very soon), but it just allows multiple Tasks on the same executor to load the Dataset in parallel. It doesn't constrain OpenMP thread usage directly.
> 
> I have met the same problems, and according to your reply I reset my cluster resource 「200cores = 50executor x 4 core」to 「200cores = 200executor x 1core」greatly accelerate the traning time. But now I get stuck in other things #1595. And I'm testing update the version from 0.10.0 to 0.10.2.

Update version to 0.10.2 solved the error mentioned in [#1595](https://github.com/microsoft/SynapseML/issues/1595), .

**Comment by sinnfashen** [comment]
Hi, I'm confused by how useSingleDataSet works here and was wondering if anyone might provide any clarification. 

For example, if we are running a 10 executors * 4 cores cluster, and we set useSingleDataSet to be true, will all cores be used for training after data preparation is done, or just 1 core from each 10 executors = 10 cores? I initially thought it's the former one, but based on @svotaw's explanation and @shexuan's example it seems to be the latter one? 

- If only one core per executor will be used, why would useSingleDataSet improve performance by using less computation resource? And in this case, is it recommended to use on core executors?
- If not, why for @shexuan's case the performance was greatly improved? I also observed from Databricks metrics page that when using SingleDatasetMode the CPU usage was not even high when using 16 cores executors.

In general, when using SingleDatasetMode, what's the suggested configuration?

**Comment by svotaw** [comment]
@sinnfashen 
Basically, SynapseML turns Spark partition Tasks into LightGBM "machines" (the unit of execution in native LightGBM code, each with its own Dataset). The parameter "useSingleDataset" controls how that conversion happens.

If useSingleDataset=false, we will initialize LightGBM to run with numTasks "machines" (i.e., one per partition). Usually this will mean more than 1 LightGBM "machine" per executor. LightGBM threads on the same physical node will share cores, and use higher-latency network protocols to communicate between ones which are actually on the same physical Spark node.
If useSingleDataset=true, then we will initialize LightGBM to run with just 1 "machine" per executor, generally resulting in a smaller number of machines performing training.

Sorry, when I was using "task" in previous comments, I meant Spark Task, not any reference to process threads or cores. We don't control "core" usage. That's up to the OS and native LightGBM library. If a LightGBM training "machine" is run on a node with 4 cores, it will likely try and use all of them (because of OpenMP integration). Multiple "machines" on the same executor will share cores. More cores should always speed up training.

Note: unless you are using some custom configuration, usually 1 executor == 1 Spark node.

Example: 10 executors available with 4 cores/node (40 cores total); 50 data partitions
useSingleDataset=false: There will be 50 LightGBM "machines" created, 1 per partition (and run on the associated Spark Task). With only 10 executors (i.e., nodes), there will be 5 "machines" per executor and hence 5 LightGBM training threads sharing 4 cores for that executor (slower execution). Also, slower network protocols will be used between 50 machines, despite some of them being on the same physical node.
useSingleDataset=true: There will be only 10 LightGBM "machines" created, 1 per executor. On each executor, 5 partitions will be merged into 1 dataset (and hence 4 Spark Tasks will exit once Dataset is created, leaving 1 to be the "machine"). With only 1 LightGBM machines per executor, training on that machine can use all 4 cores. There are also only 10 training machines running, reducing overall network traffic to a minimum.

So, depending on how the cores/partitions/nodes are configured and related, useSingleDataset=true will USUALLY be faster. You can come up with some configurations where is doesn't help or is even slower.

As far as your example, it's hard to comment without knowing the number of partitions you are using.

I hope that helps. Feel free to ask more questions.

**Comment by ranadeepsingh** [comment]
Closing as answered. The thread contains detailed explanations from maintainers regarding how useSingleDataset interacts with Spark executors and OpenMP threading in LightGBM, and why task slots cannot be locked during fitting.
